hazelcast/src/hazelcast/client/sql.cpp

```diff
@@ -0,0 +1,918 @@
+  , code_(code)
+  , suggestion_(std::move(suggestion))
+{
+}
+
+const boost::uuids::uuid&
+hazelcast_sql_exception::originating_member_id() const
+{
+    return originating_member_id_;
+}
+
+int32_t
+hazelcast_sql_exception::code() const
+{
+    return code_;
+}
+
+const boost::optional<std::string>&
+hazelcast_sql_exception::suggestion() const
+{
+    return suggestion_;
+}
+
+namespace impl {
+
+void
+query_utils::throw_public_exception(std::exception_ptr exc,
+                                    boost::uuids::uuid id)
+{
+    try {
+        std::rethrow_exception(exc);
+    } catch (hazelcast_sql_exception& e) {
+        throw;
+    } catch (exception::query& e) {
+        auto originating_member_id = e.originating_member_uuid();
+        if (originating_member_id.is_nil()) {
+            originating_member_id = id;
+        }
+
+        throw hazelcast_sql_exception("query_utils::throw_public_exception",
+                                      originating_member_id,
+                                      e.code(),
+                                      e.get_message(),
+                                      e.suggestion(),
+                                      exc);
+    } catch (exception::iexception& ie) {
+        throw hazelcast_sql_exception(
+          "query_utils::throw_public_exception",
+          id,
+          static_cast<int32_t>(sql_error_code::GENERIC),
+          ie.get_message(),
+          boost::none,
+          exc);
+    }
+}
+
+boost::optional<member>
+query_utils::member_of_same_larger_version_group(
+  const std::vector<member>& members)
+{
+    // The members should have at most 2 different version (ignoring the patch
+    // version). Find a random member from the larger same-version group.
+
+    // we don't use 2-element array to save on litter
+    boost::optional<member::version> version0;
+    boost::optional<member::version> version1;
+    size_t count0 = 0;
+    size_t count1 = 0;
+
+    for (const auto& m : members) {
+        if (m.is_lite_member()) {
+            continue;
+        }
+        auto v = m.get_version();
+        if (!version0 || *version0 == v) {
+            version0 = v;
+            ++count0;
+        } else if (!version1 || *version1 == v) {
+            version1 = v;
+            ++count1;
+        } else {
+            throw exception::runtime(
+              "query_utils::member_of_same_larger_version_group",
+              (boost::format(
+                 "More than 2 distinct member versions found: %1% , %2%") %
+               version0 % version1)
+                .str());
+        }
+    }
+
+    assert(count1 == 0 || count0 > 0);
+
+    // no data members
+    if (count0 == 0) {
+        return boost::none;
+    }
+
+    size_t count;
+    member::version version;
+    if (count0 > count1 || (count0 == count1 && *version0 > *version1)) {
+        count = count0;
+        version = *version0;
+    } else {
+        count = count1;
+        version = *version1;
+    }
```

Comment:
You may consider computing and caching the `version` when the member list is received That makes the code different from the Java client but it is more performant and the end result should be the same.

Reply:
This can be done in Java impl. as well. I would keep the code here but I opened an issue [here](https://github.com/hazelcast/hazelcast-cpp-client/issues/1023) so that we can change both Java and here.